`used-before-assignment` with `:=` in `dict` node

### Bug description

https://github.com/PyCQA/pylint/issues/4828#issuecomment-1236297125

```py
def func(): ...
y = {
    0: x if (x := func()) else "",
}
```

### Configuration

_No response_

### Command used

```shell
pylint test.py
```


### Pylint output

```shell
test.py:3:7: E0601: Using variable 'x' before assignment (used-before-assignment)
```


### Expected behavior

_No error_

### Pylint version

```shell
pylint 2.16.0b1
astroid 2.14.0-dev0
Python 3.11.1
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8125.bugfix
+++ b/doc/whatsnew/fragments/8125.bugfix
@@ -1,0 +1,4 @@
+Fix ``used-before-assignment`` false positive when the walrus operator
+is used with a ternary operator in dictionary key/value initialization.
+
+Closes #8125

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2262,6 +2262,8 @@ class VariablesChecker(BaseChecker):
             return True
         if isinstance(value, nodes.Lambda) and isinstance(value.body, nodes.IfExp):
             return True
+        if isinstance(value, nodes.Dict) and any(isinstance(item[0], nodes.IfExp) or isinstance(item[1], nodes.IfExp) for item in value.items):
+            return True
         if not isinstance(value, nodes.Call):
             return False
         return any(

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2262,7 +2262,10 @@ class VariablesChecker(BaseChecker):
             return True
         if isinstance(value, nodes.Lambda) and isinstance(value.body, nodes.IfExp):
             return True
-        if isinstance(value, nodes.Dict) and any(isinstance(item[0], nodes.IfExp) or isinstance(item[1], nodes.IfExp) for item in value.items):
+        if isinstance(value, nodes.Dict) and any(
+            isinstance(item[0], nodes.IfExp) or isinstance(item[1], nodes.IfExp)
+            for item in value.items
+        ):
             return True
         if not isinstance(value, nodes.Call):
             return False

--- a/tests/functional/u/used/used_before_assignment_ternary.py
+++ b/tests/functional/u/used/used_before_assignment_ternary.py
@@ -29,6 +29,19 @@ def function_call_keyword_valid():
     var = foo(x=a if (a:='1') else '', y='', z='')
     var = foo(x='', y=foo(x='', y='', z=b if (b:='1') else ''), z='')
 
+def dictionary_items_valid():
+    """assignment as dictionary keys/values"""
+    var = {
+        0: w if (w:=input()) else "",
+    }
+    var = {
+        x if (x:=input()) else "": 0,
+    }
+    var = {
+        0: y if (y:=input()) else "",
+        z if (z:=input()) else "": 0,
+    }
+
 def complex_valid():
     """assignment within complex call expression"""
     var = str(bar(bar(a if (a:=1) else 0))).lower().upper()

